source/extensions/filters/http/json_to_metadata/filter.cc

```diff
@@ -0,0 +1,376 @@
+    } else {
+      return double_result.status();
+    }
+    break;
+  case envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata::STRING:
+    std::string str = absl::visit(JsonValueToStringConverter(), std::move(result.value()));
+    if (str.size() > MAX_PAYLOAD_VALUE_LEN) {
+      return absl::InvalidArgumentError(
+          fmt::format("metadata value is too long. value.length: {}", str.size()));
+    }
+
+    // Note that this applies to on_present by not adding any metadata.
+    if (str.empty()) {
+      ENVOY_LOG(debug, "value is empty, not adding metadata. key: {}", on_present_keyval.key());
+      return absl::OkStatus();
+    }
+
+    applyKeyValue(std::move(str), on_present_keyval, struct_map);
+    break;
+  }
+  return absl::OkStatus();
+}
+
+void Filter::processBody(const Buffer::Instance* body, const Rules& rules, bool& reported_flag,
+                         Stats::Counter& success, Stats::Counter& no_body,
+                         Stats::Counter& non_json) {
+  if (!body) {
+    handleAllOnMissing(rules, is_request_reported_);
+    no_body.inc();
+    return;
+  }
+
+  absl::StatusOr<Json::ObjectSharedPtr> result =
+      Json::Factory::loadFromStringNoThrow(body->toString());
+  if (!result.ok()) {
+    ENVOY_LOG(debug, result.status().message());
+    non_json.inc();
+    handleAllOnError(rules, reported_flag);
+    return;
+  }
+
+  Json::ObjectSharedPtr body_json = std::move(result.value());
+  StructMap struct_map;
+  for (const auto& rule : rules) {
+    const auto& keys = rule.keys();
+    Json::ObjectSharedPtr node = body_json;
+    bool on_missing = false;
+    for (unsigned long i = 0; i < keys.size() - 1; i++) {
+      absl::StatusOr<Json::ObjectSharedPtr> next_node_result = node->getObjectNoThrow(keys[i]);
+      if (!next_node_result.ok()) {
+        ENVOY_LOG(warn, result.status().message());
+        handleOnMissing(rule, struct_map);
+        on_missing = true;
+        break;
+      }
+      node = std::move(next_node_result.value());
+    }
+    if (on_missing) {
+      continue;
+    }
+    absl::Status result = handleOnPresent(std::move(node), keys.back(), rule, struct_map);
+    if (!result.ok()) {
+      ENVOY_LOG(warn, fmt::format("{} key: {}", result.message(), keys.back()));
+      handleOnMissing(rule, struct_map);
+    }
+  }
+  success.inc();
+
+  finalizeDynamicMetadata(*decoder_callbacks_, struct_map, reported_flag);
+}
+
+void Filter::processRequestBody() {
+  processBody(decoder_callbacks_->decodingBuffer(), config_->requestRules(), is_request_reported_,
+              config_->stats().rq_success_, config_->stats().rq_no_body_,
+              config_->stats().rq_invalid_json_body_);
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  ASSERT(config_->doRequest());
+  if (!config_->requestContentTypeAllowed(headers.getContentTypeValue())) {
+    is_request_reported_ = true;
```

Comment:
what exactly is a reported request?

Reply:
hmmm. it means the request body is parsed and set the corresponding metadata. should be `is_request_handled_` ?

Reply:
I'll use `request_processing_finished_` if you don't mind